src/plugins/discover/public/application/main/utils/use_discover_state.ts

```diff
@@ -231,6 +176,49 @@ export function useDiscoverState({
     }
   }, [initialFetchStatus, refetch$, indexPattern, savedSearch.id]);
 
+  /**
+   * Track state changes that should trigger a fetch
+   */
+  useEffect(() => {
+    const unsubscribe = appStateContainer.subscribe(async (nextState) => {
+      const { hideChart, interval, sort, index } = state;
+      // chart was hidden, now it should be displayed, so data is needed
+      const chartDisplayChanged = nextState.hideChart !== hideChart && hideChart;
+      const chartIntervalChanged = nextState.interval !== interval;
+      const docTableSortChanged = !isEqual(nextState.sort, sort);
+      const indexPatternChanged = !isEqual(nextState.index, index);
+      // NOTE: this is also called when navigating from discover app to context app
+      if (nextState.index && indexPatternChanged) {
+        /**
+         *  Without resetting the fetch state, e.g. a time column would be displayed when switching
+         *  from a index pattern without to a index pattern with time filter for a brief moment
+         *  That's because appState is updated before savedSearchData$
+         *  The following line of code catches this, but should be improved
+         */
+        const nextIndexPattern = await loadIndexPattern(nextState.index, indexPatterns, config);
+        savedSearch.searchSource.setField('index', nextIndexPattern.loaded);
+
+        reset();
+      }
+
+      if (chartDisplayChanged || chartIntervalChanged || docTableSortChanged) {
+        refetch$.next();
+      }
+      setState(nextState);
+    });
+    return () => unsubscribe();
+  }, [
+    config,
+    indexPatterns,
+    appStateContainer,
+    setState,
+    state,
+    refetch$,
+    data$,
+    reset,
+    savedSearch.searchSource,
+  ]);
+
```

Comment:
Since we need to backport this, I'd suggest keeping the number of changes small. In this case just the position in the file changed, could we keep the position? Since in this files there are plenty of changes, and just a small number are relevant to the fix